### ISSUE TITLE
Problems building Hercules on Raspberry Pi for ARM / AARCH64

### ISSUE THREAD

**Issue by bert269** [SDL-Hercules-390/hyperion#334]
_**NOTE:** This issue may be related to issue #352 as well, since they both deal with building Hercules._
___
Has anyone tried to run Hyperion (Hercules) on a Raspberry Pi?

Well, in case you did not know, we have 4-core, 1.5GHz processor with 8GB or RAM. This is running Ubuntu Mate 20.10 on the rpi.

Crazy? Sure.

Everything goes fine with the `bldlvlck` and the configure commands. The problem happens on the make statement. I get these errors and have _NO IDEA_ where to go from here. It seems that we need something for ARM or AARCH64 that it does not like?

```
*** Warning: Linking the shared library libhercu.la against the
*** static library ./crypto/lib/aarch64/libcrypto64.a is not portable!

*** Warning: Linking the shared library libhercu.la against the
*** static library ./decNumber/lib/aarch64/libdecNumber64.a is not portable!

*** Warning: Linking the shared library libhercu.la against the
*** static library ./SoftFloat/lib/aarch64/libSoftFloat64.a is not portable!

*** Warning: Linking the shared library libhercu.la against the
*** static library ./telnet/lib/aarch64/libtelnet64.a is not portable!
gcc: error: ./crypto/lib/aarch64/libcrypto64.a: No such file or directory
gcc: error: ./decNumber/lib/aarch64/libdecNumber64.a: No such file or directory
gcc: error: ./SoftFloat/lib/aarch64/libSoftFloat64.a: No such file or directory
gcc: error: ./telnet/lib/aarch64/libtelnet64.a: No such file or directory
make[2]: *** [Makefile:2082: libhercu.la] Error 1
make[2]: Leaving directory '/home/pi/Downloads/hyperion-master'
make[1]: *** [Makefile:2489: all-recursive] Error 1
make[1]: Leaving directory '/home/pi/Downloads/hyperion-master'
make: *** [Makefile:1829: all] Error 2
```

Any ideas? I apologize if this error have been reported before - just call me crazy and I will go search for the solution.

Thanks

Bert

**Comment by wrljet** [comment]
I can assist you with that. 

Can you post the output from  `uname -a`

And what build command/steps did you use?

Bill

**Comment by Fish-Git** [comment]
Bert (@bert269),

Are you still having this problem? Did you see the offer to help you from Bill Lewis? He asked you some questions which you still have not answered yet. Please login to your GitHub account and go to https://github.com/SDL-Hercules-390/hyperion/issues/334 to respond to Bill's questions.

Additionally, I have my own question to ask you:

* Did you read and carefully follow the instructions documented on our web page that explains how to build SDL Hyperion? The URL is: https://sdl-hercules-390.github.io/html/hercinst.html#instnixmac. Step 5 is especially important! Did you do complete 5 before attempting to build Hercules?

**Comment by bert269** [comment]
> I can assist you with that.
> 
> Building on Pi does work for both 32- and 64-bit, but I haven't tried a current Ubuntu Mint. I will set one up here.
> 
> Can you post the output from `uname -a`
> 
> And what exact build command/steps did you use?
> 
> Bill

Bill, Sorry it took me a while to get back to you. Here is the output from uname -a 👍 
**Linux piTest 5.4.79-v7l+ #1373 SMP Mon Nov 23 13:27:40 GMT 2020 armv7l GNU/Linux**

I just setup a brand new installation and I am going to try to re-install and configure Hyperion. Also, from what I read from Fish, I will look into re-doing the external packages. I have to admit, I have not done what. I will report back on what I encounter.

Thanks for your help.

**Comment by wrljet** [comment]
I think you should give my hercules-helper scripts a try.
It completely automates the entire process on Linux.

I will post instructions here a bit later after I retest it all myself on a 32-bit Pi.

Bill

**Comment by bert269** [comment]
Bill
not a problem - I will wait till I see your response.
Appreciate the help!

**Comment by wrljet** [comment]
Please also post the output from `uname -m` and `cat /etc/os-release`

**Comment by bert269** [comment]
**uname -m**
armv71

**cat /etc/os-release**
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
NAME="Raspbian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

**Comment by wrljet** [comment]
OK, so this is different than your original post, which mentioned Ubuntu Mate 20.10.

Which is cool.  My aim is to make everything build smoothly.

**Comment by bert269** [comment]
Yes, that is correct, sorry. I first tried Ubuntu 20.10 and after I could not make it to work, I reverted back to RaspianOs this morning. But if you prefer, I can always re-install ubuntu also - it's just a "sdcard-burn" away.

**Comment by wrljet** [comment]
No problem.
Ubuntu 20.10 64-bit is fully tested in my scripts now.  Works.
I'm going to swap SD cards and boot up the RPI OS 32-bit shortly.

**Comment by bert269** [comment]
Awesome, thanks for the update and your help. I may not get to it today as
I have to go into the office a little later...maintenance.
Will try and get to it tomorrow after work. Thank you.

**Comment by wrljet** [comment]
I might get it done today. :-)

**Comment by wrljet** [comment]
The helper scripts may be obtained with:

`git clone https://github.com/wrljet/hercules-helper.git`

or, if you don't have git, or simply prefer:

`wget https://github.com/wrljet/hercules-helper/archive/master.zip`

You will need to `chmod +x` the three Bash scripts.

The default README at the bottom of the GitHub repo page will display some instructions.
You should check that out before doing anything else.

`https://github.com/wrljet/hercules-helper`

But basically, this is all you need to do:

```
$ cd ~
$ git clone https://github.com/wrljet/hercules-helper.git
$ chmod +x ~/hercules-helper/*.sh

$ ~/hercules-helper/hyperion-prepare.sh --verbose

$ mkdir herctest && cd herctest
$ ~/hercules-helper/hyperion-buildall.sh --verbose --prompts --install 2>&1 | tee ./hyperion-buildall.log
```

If packages need to be installed you may be asked to supply your sudo password.

You will be prompted a number of times between the major steps, to give you a chance
to see the results of the last step, and to clue you into what will be happening next.

Hercules will be "installed" if you include the --install option as shown above, into `~/herctest/herc4x`

To set the required environment variables after installation, run
    `source ~/herctest/hercules-setvars.sh`

If anything seems to go wrong, please stop and ask questions at that point.
Your repair attempts may destroy evidence that would be useful in improving this process for others.

Thanks,
Bill

**Comment by bert269** [comment]
Thank you Bill. I am downloading it now from Git and will let you know upon
success.
Thanks for your time and help.

**Comment by bert269** [comment]
Thanks Bill

During the run of Hercules, I got one message. Now I did not start it up
with a configuration file yet - I was trying to see what happened and it
seemed to have worked, could not find the config file and ended.

But I saw this message - not sure if it is related to the fact that I did
not specify the config file:
HHC00100I Thread id b6fbe010, prio 5, name 'impl_thread' started
HHC00100I Thread id b64cd440, prio 4, name 'logger_thread' started
HHC01413I Hercules version 4.4.9999.0-SDL-g346bcfe2-modified (4.4.9999.0)
HHC01414I (C) Copyright 1999-2020 by Roger Bowler, Jan Jaeger, and others
HHC01417I ** The SoftDevLabs version of Hercules **
HHC01415I Build date: Dec 14 2020 at 14:53:36
HHC01417I Built with: GCC 8.3.0
HHC01417I Build type: GNU/Linux armv7l host architecture build
HHC01417I Modes: S/370 ESA/390 z/Arch
HHC01417I Max CPU Engines: 64
HHC01417I Using   shared libraries
HHC01417I Using   setresuid() for setting privileges
HHC01417I Using   POSIX threads Threading Model
HHC01417I Using   Error-Checking Mutex Locking Model
HHC01417I With    Shared Devices support
HHC01417I With    Dynamic loading support
HHC01417I With    External GUI support
HHC01417I With    IPV6 support
HHC01417I With    HTTP Server support
HHC01417I With    sqrtl support
HHC01417I With    Signal handling
HHC01417I With    Watchdog monitoring
HHC01417I With    CCKD BZIP2 support
HHC01417I With    HET BZIP2 support
HHC01417I With    ZLIB support
HHC01417I With    Regular Expressions support
HHC01417I Without Object REXX support
HHC01417I With    Regina REXX support
HHC01417I With    Automatic Operator support
HHC01417I Without National Language Support
HHC01417I With    CCKD64 Support
HHC01417I With    Transactional-Execution Facility support
HHC01417I With    "Optimized" instructions
HHC01417I Machine dependent assists: cmpxchg1 cmpxchg4 cmpxchg8
hatomics=atomicIntrinsics
HHC01417I Running on: piTest (Linux-5.4.79-v7l+ armv7l) MP=4
HHC01417I Built with crypto external package version 1.0.0.43-gbb06b3a
HHC01417I Built with decNumber external package version 3.68.0.97-g38ad260
HHC01417I Built with SoftFloat external package version 3.5.0.100-gacc8140
HHC01417I Built with telnet external package version 1.0.0.57-g70874fe
HHC00018W Hercules is NOT running in elevated mode
HHC00007I Previous message from function 'impl' at impl.c(1021)
HHC00150I Crypto module loaded (C) Copyright 2003-2016 by Bernard van der
Helm
HHC00151I Activated facility: Message Security Assist
HHC00151I Activated facility: Message Security Assist Extension 1, 2, 3 and
4
*HHC17531W REXX(Regina) dlopen 'libregina.so' failed: libregina.so: cannot
open shared object file: No such file or directory*
HHC00007I Previous message from function 'ReginaLoad' at hRexxapi.c(1314)
HHC17511E REXX() Could not enable either Rexx package
HHC00007I Previous message from function 'Enable' at hRexx.c(515)
HHC17511E REXX() Could not enable default Rexx package
HHC00007I Previous message from function 'InitializeRexx' at hRexx.c(262)
HHC17500I REXX() Mode            : Command
HHC17500I REXX() MsgLevel        : Off
HHC17500I REXX() MsgPrefix       : Off
HHC17500I REXX() ErrPrefix       : Off
HHC17500I REXX() Resolver        : On
HHC17500I REXX() SysPath    ( 8) : On
HHC17500I REXX() RexxPath   ( 0) :
HHC17500I REXX() Extensions ( 8) : .REXX:.rexx:.REX:.rex:.CMD:.cmd:.RX:.rx
HHC00111I Thread CPU Time IS available (_POSIX_THREAD_CPUTIME=0)
HHC00100I Thread id b5cfc440, prio 2, name 'Processor CP00' started
HHC00100I Thread id b59ff440, prio 7, name 'timer_thread' started
HHC00811I Processor CP00: architecture mode z/Arch
*HHC01432S Config file[1] : error in function fopen(): No such file or
directory*
HHC00007I Previous message from function 'process_config' at script.c(217)
HHC01420I Begin Hercules shutdown
HHC01423I Calling termination routines
HHC00101I Thread id b5cfc440, prio 2, name 'Processor CP00' ended
HHC01427I Main storage released
HHC01427I Expanded storage released
HHC01422I Configuration released
HHC01424I All termination routines complete
HHC01425I Hercules shutdown complete
HHC01412I Hercules terminated
HHC00101I Thread id b59ff440, prio 7, name 'timer_thread' ended

Also , I did save all the installation .log files - if there are anything
you want me to confirm in those.

Appreciated.
Bert

>

**Comment by wrljet** [comment]
That is how Hercules reacts, if you run it without any config file.

If you ran the last command I mentioned: `source ~/herctest/hercules-setvars.sh` it should have found Regina REXX, though.

Why don't you upload the `hyperion-buildall.log` that was left behind.
I'd like to see exactly what it did through the process.

Bill

**Comment by bert269** [comment]
Bill

Thanks no problem. I did run the hercules-servars.sh script. Will upload
the log file.

Thank you

**Comment by 00003481** [comment]
sorry for the late question Bill
do the floating point tests (BFP-001 to BFP-022) pass on the arm

enrico

**Comment by wrljet** [comment]
> sorry for the late question Bill
> do the floating point tests (BFP-001 to BFP-022) pass on the arm
> 
> enrico

The floating point tests do pass on the various Pi editions.

**Comment by 00003481** [comment]
That You  Bill,

from the  SoftFloat docs...
> Added specialization code modeling ARM floating-point, conforming to VFPv2 or later.

a quick and dirty search returned that the raspberry pi 4 has the VFPv2 level
so I wonder if it is worth investigating the use of the specialised  ARM functions 

 as soon as my PI arrives 
I will try to build softFloat with the ARM things  and see what happens 

and at first  glance I noticed some difference in some constants used

if You are interested as soon as I have something working 
I will keep You posted with the links to my repositories 

cheers

enrico

**Comment by wrljet** [comment]
Are we done with this issue?

**Comment by Fish-Git** [comment]
> Are we done with this issue?

You tell ME, Bill! &nbsp; `:)`

Are your helper scripts ready to be committed yet?

Does our documentation maybe need updating? Do we maybe need to add detailed instructions specifically for Raspberry Pi for ARM / AARCH64?

I too would like to close this issue! If you feel it's been resolved, then please close it. Otherwise I'm marking it as "Waiting to close" and will close it myself in a few days if there is no further progress. _(No sense in keeping issues open if they've reached a dead end!)_

**Comment by wrljet** [comment]
Fish,

> > Are we done with this issue?
> 
> You tell ME, Bill!   `:)`

As far as I'm concerned the stuff about the SoftFloat is a red herring.
Nothing is needed there.


> Are your helper scripts ready to be committed yet?

They are on my Github, but still probably "under construction" depending.
Just today I started working on adding FreeBSD for Pi to it after reading Bernd's message.

 
> Does our documentation maybe need updating? Do we maybe need to add detailed instructions specifically for Raspberry Pi for ARM / AARCH64?

A problem is the _details_ are somewhat different for different *nix flavors.

 
> I too would like to close this issue! If you feel it's been resolved, then please close it. Otherwise I'm marking it as "Waiting to close" and will close it myself in a few days if there is no further progress. _(No sense in keeping issues open if they've reached a dead end!)_

Not a dead end.  Still improving it.

Bill

**Comment by wrljet** [comment]
For now, we *could* add instructions pointing people to my hercules-helper and simply list the known Pi OSes where it's believed to work.

**Comment by Fish-Git** [comment]
> For now, we _could_ add instructions pointing people to my hercules-helper and simply list the known Pi OSes where it's believed to work.

So... How should we proceed? Do you want to update (commit the needed changes to) our build instructions documentation to do that? (point to your [still-under-development] helper scripts) and then close this issue? Or do you want to keep this issue open until such time as your helper scripts are finished?

**Comment by BerndHKG** [comment]
Hi Fish,

so I tried building Hyperion using build instructions from the correct URL (i.e. SDL's :-| ). First step was to build the external packages and here I ran into the same issue as Sunil when he tried it on his M1 Mac, i.e. extpkgs.sh would build 32-bit code instead of 64-bit code. The fix/hack was fairly simple as uname -m returns _arm64_:
```
get_default_cpu()
{
  case "$(uname -m)" in
  
    arm64*)
      default_cpu="aarch64"
      ;;
  
    aarch64*)
      default_cpu="aarch64"
      ;;
```
after this I got 64-bit code.

When I run configure, I get stuck at
```
configure: Using External Packages override directory /home/hercules/source/extpkgs/. 
configure: NOTICE: sizeof(__int128_t) >= 16; max CPU engines = 128 
checking whether getopt wrapper kludge is necessary... no
ERROR: external-gui requires loadable modules (shared libraries = --enable-shared) 
configure: error: Please correct the above error(s) and try again
``` 
So far I couldn't figure out what I should do, `--enable-shared` doesn't change anything.

Bernd

**Comment by Fish-Git** [comment]
> When I run configure, I get stuck at:
> 
> ```
> ERROR: external-gui requires loadable modules (shared libraries = --enable-shared) 
> configure: error: Please correct the above error(s) and try again
> ```
> 
> So far I couldn't figure out what I should do, `--enable-shared` doesn't change anything.

I've never seen that before!

Maybe something went wrong earlier in your configure to cause the problem you are seeing? Please post your full ./configure output so we can see it. (What `./configure` options did you specify by the way? Hopefully Bill will be able to shed some light on the subject?)

**Comment by wrljet** [comment]
Yes, I've gotten past that (and some other issues) and have it building on FreeBSD 12.2 on a Pi 3B+.
Hercules "runs" but I haven't run the tests yet.

I'll fill you in on the specifics this afternoon.

Bill

**Comment by mgrossmann** [comment]
Building latest SDL hyperion on RPI4 (aarch64) works fine for me. Did it "by hand" as well as using the helper script from Bill. 

No problem at all.

**Comment by wrljet** [comment]
re: building on FreeBSD 12.2 on a Pi 3B+

Bernd,

OK, it took 4 days of fiddling with it, on and off.  Not just the afternoon I had hoped for. :-)
(thankfully my cussing skills are still in good shape.)

But it's working!
All but the `mainsize` test, which is know to fail on *nix with low RAM, which I've therefore ignored.

I need to diff everything I've done and get it into shape for you to try out.

Bill

**Comment by BerndHKG** [comment]
Great work Bill! I'm looking forward to trying it out.

**Comment by wrljet** [comment]
Bernd,

Are you running a GUI window interface on your FreeBSD Pi 3?
Or just the plain text system?

Bill

**Comment by BerndHKG** [comment]
Hi Bill,

no GUI, just terminal windows. Keep it at least a little simple :-)

Bernd

**Comment by wrljet** [comment]
Bernd,

In FreeBSD, is there a way to scroll back in the Alt+N terminal screens?

Bill

**Comment by BerndHKG** [comment]
Hi Bill,

I just ssh into my Pi and the MacOS terminal scrolls very nicely.

Bernd

**Comment by wrljet** [comment]
Bernd, let's give something a try on FreeBSD 12.2 for Raspberry Pi 3B+

**Please read everything before doing anything.**

First off, I try my best to not wreck anything.  But you should back up your SD card before trying this out.

If and when things diverge from what I'm describing here, it's probably best to stop and tell me
what's happened.  Otherwise evidence of the problem could be lost that would help us improve this
process for others.

I have tested this with both the default Clang compiler that comes with FreeBSD 12.2, and gcc 8.

I assume you have Git already installed.  If not, you can run:

`sudo pkg install git`

This process requires Bash.  When I installed the FreeBSD 12.2 ISO image for the Pi, the default
userid, freebsd, came up using csh.  I installed Bash (and also edited `/etc/passwd` to change the login
to use it in place of csh).

Run this command from your home dir, to pull down my Hercules Helper project.
It will checkout the "freebsd" branch.

```
cd ~
git clone -b freebsd https://github.com/wrljet/hercules-helper.git
```

(( Currently this also includes all the Windows specific stuff, which we aren't using here.
In future I'm going to split that out to not clog up the Linux version.  For now you can just ignore it. ))

The two main files are `hyperion-buildall.sh` and `hercules-helper.conf`.
The config file in this freebsd branch is set up to point to my forked/branched copies of the SDL-Hercules repos,
with my FreeBSD modifications.

Create a directory in which to run all this and cd to it.  For example:

`mkdir ~/herctest && cd ~/herctest`

If you want to see all the options, you can run:

`~/hercules-helper/hyperion-buildall.sh --help`

Then we can run:

`~/hercules-helper/hyperion-buildall.sh -v -p`

`-v` means Verbose, `-p` means Prompt

You will see a huge amount of progress output, which is also automatically logged to a file.
The log filenames are timestamped and will not be overwritten by subsequent runs.

Before doing each major step of the process, you will see a prompt announcing what it is about to do, and
asked to press return to continue.

Some of the steps take a while to get warmed up, so be patient and don't hit Enter more than once
as that might be used to answer the next prompt before you are ready.

The 'mainsize' test will be skipped because the Pi 3 does not have enough RAM, and FreeBSD
will kill the process, canceling the entire thing.

You may be asked for your sudo password along the way.

The major steps are:

- general system detection
- install required system packages
- build Regina REXX (used for the tests, only built if no REXX is found)
- 'git clone' the Hercules repos
- run 'util/bldlvlck'
- build Hercules external packages
- run 'autogen'
- run 'configure'
- run 'make clean'
- run 'make\'
- run 'make check'
- run 'make install'
- create script to set environment variables
- modify .bashrc to set environment variables

Bill

**Comment by BerndHKG** [comment]
Hi Bill,

it works! And with referencing hercules.cnf Hyperion also starts up and that even pretty fast. I also can connect using c3270. Now I can start to figure out how to install an OS and what to do with my Raspberry S/390 ... or is it even a z314? 

Thanks for doing all this work to get Hyperion running on the RPi and even under FreeBSD :-). If there is more testing to do I'd gladly help.

Best regards
Bernd

**Comment by wrljet** [comment]
Bernd,

That is wonderful news!

I'd like to inspect the log it created, just to ensure nothing seems unexpected.
In the `herctest` directory, you should have a big log file.  Please zip and email that to me, `bill@wrljet.com`.

Thanks,
Bill

**Comment by wrljet** [comment]
For what it's worth, the same changes work for FreeBSD 13 RC5 on Raspberry Pi 4B.
And FreeBSD 12.2 on x86_64.

It's been checked into GitHub on SDL-Hercules-390, and my Hercules-Helper.

**Comment by wrljet** [comment]
Just as a quickie, can you run this and then try Hercules again:

`sudo ldconfig`

Bill

**Comment by wrljet** [comment]
Allard,

Your hercules.log still contains this msg:

`HHC17531W REXX(Regina) dlopen 'libregina.so' failed: libregina.so: cannot open shared object file: No such file or directory`

Is that the correct / current log file?

Bill

**Comment by AllardKrings** [comment]
Hello Bill,

you are correct! I overlooked that error, sorry.

I did the test again this morning:

SUDO LDCONFIG

and then 

sudo -i /home/ubuntu/herctest/herc4x/bin/hercules -f /share/hercules.cnf -o /share/hercules.log

The  remaining errors are:

06:10:19 HHC17531W REXX(Regina) dlopen 'libregina.so' failed: libregina.so: cannot open shared object file: No such file or directory
06:10:19 HHC00007I Previous message from function 'ReginaLoad' at hRexxapi.c(1339)
06:10:19 HHC17511E REXX() Could not enable either Rexx package
06:10:19 HHC00007I Previous message from function 'Enable' at hRexx.c(515)
06:10:19 HHC17511E REXX() Could not enable default Rexx package
06:10:19 HHC00007I Previous message from function 'InitializeRexx' at hRexx.c(262)

And:

06:10:19 HHC00136E Error in function execlp(): No such file or directory
06:10:19 HHC00007I Previous message from function 'IFC_IOCtl' at tuntap.c(1118)

Thanks for your patience and help!

Kind regards Allard

**Comment by wrljet** [comment]
OK, this has been reproduced here on a Pi 4 with a fresh Ubuntu 21.04.
Problem found and fixed / workaround.
Let me get my stuff sorted out and I'll post what to do.

Bill

**Comment by wrljet** [comment]
Allard,

Both the symptoms you are seeing, the REXX and the tuntap/execlp errors are caused by the similar problems.
Which is simply that when you run as sudo/root, the environment and the rules of the loader are different.
This is for security reasons.  The exact behavior also varies by Linux distribution and how old it is.

`libregina.so` cannot be found because LD_LIBRARY_PATH is ignored under sudo.

The `hercifc` binary (for your networking) cannot be found due to an interaction between setcapabilities and sudo.

---

The REXX issue is addressed by changes I made to Hercules-Helper a few days ago.
It no longer installs the locally built Regina inside the `herctest` directory.  But instead installs
it in the system default locations.  This was done unrelated to your issue, but turns out was a good idea. :-)
With this change, Regina is no longer needs to be mentioned in `$PATH` or `$LD_LIBRARY_PATH`.

Also I found a problem with the autoconf scripts in the Regina package causing it to not build correctly on 64-bit Pi.
They are simply too old.  I have worked around this by patching `configure` during the build.

---

Your networking can be solved by *not* running Hercules as root with sudo.
This is causing Hercules to not be able to start `hercifc`.
Hercules is installed with `setcap`, so running as root is not required, and the complex rules are causing it to not be found.

If for some reason you *really* want to run it with sudo you can set an `HERCULES_IFC=` environment variable so Hercules knows
where to find it.

For example:
`sudo -i HERCULES_IFC=/home/ubuntu/herctest/herc4x/bin/hercifc /home/ubuntu/herctest/herc4x/bin/hercules -f ...`

---

So...

I recommend you throw away your existing Hercules-Helper directory and do a new fresh git clone.

`git clone https://github.com/wrljet/hercules-helper.git`

Then rerun `hyperion-buildall.sh` and you should be all set!

(fingers crossed)
Bill

P.S.  Please send me the logs, if it works or not, so I can just verify things are as I hoped.

**Comment by AllardKrings** [comment]
Hi Bill,

thanks for your help and quick response.

Apparently from the past I assumed that Hercules HAD to be run in elevated mode to make TUN0 work. 

I did a fresh git clone of hercules-helper and rebuilt the system. 

I run it without sudo -i

the I can ping the network interface on the local machine from within Hercules and can ping the Hercules ip from the local machine 
I added a gw to the interface 

sudo route add 192.168.248 gw 192.168.2.11

tun0 interface is now configured as::

tun0: flags=81<UP,POINTOPOINT,RUNNING>  mtu 1500
        inet 192.168.2.11  netmask 255.255.255.255  destination 192.168.2.248
        inet6 fe80::658f:176c:6785:7629  prefixlen 64  scopeid 0x20<link>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 18  bytes 11222 (11.2 KB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

I still cant ping the inside of Hercules from other machines, and from within Hercules I can onlu ping the local machine, so I have to do some figuring out about setting the routing correctly.

Attached you will find the buildall-lo and hercules start-up log

Kind regards!

Allard
[hyperion-buildall-2021-06-13.log](https://github.com/SDL-Hercules-390/hyperion/files/6646280/hyperion-buildall-2021-06-13.log)
[hercules.log](https://github.com/SDL-Hercules-390/hyperion/files/6646281/hercules.log)

**Comment by wrljet** [comment]
Allard,

Your `hyperion-buildall-2021-06-13.log` has errors in the configure step, and a new build was not performed.
So something's not right.  Looks like it couldn't find the `rexx*.h` file(s).
A bit more work is needed there.  I will look at it more closely a bit later.

For your specific network configuration, I suggest you ask those questions on one of the Hercules-related Groups.io lists
where the people there have the experience and know-how to assist you.

I personally know *nothing* about z/OS or its networking.
The scripts I use to get zLinux working involve `iptables` commands.

Bill

**Comment by wrljet** [comment]
Please try this command and post the output:

`find / -name 'rexx*.h' 2>/dev/null`

**Comment by AllardKrings** [comment]
Hi Bill,

here is the output you requested:

/home/ubuntu/herctest/rexx/include/rexxsaa.h
/home/ubuntu/herctest/regina-rexx-3.9.3/rexx.h
/home/ubuntu/herctest/regina-rexx-3.9.3/rexxmsg.h
/home/ubuntu/herctest/regina-rexx-3.9.3/rexxbif.h
/home/ubuntu/herctest/regina-rexx-3.9.3/rexxsaa.h

Kind regards 

Allard

**Comment by wrljet** [comment]
Somehow things appear to be in a confused state.  Not sure how it got like that.
But we will figure that out.

Please issue these commands and post the results here.

`apt search regina-rexx`
`apt search libregina`
`which -a hercules`

**Comment by AllardKrings** [comment]
Hi Bill, 

I hope I did not cause the confused state!

1: 
apt search regina-rexx
Sorting... Done
Full Text Search... Done
regina-rexx/hirsute,now 3.6-2.2build1 arm64 [installed]
  Regina REXX interpreter


2:
apt search libregina
Sorting... Done
Full Text Search... Done
libregina3/hirsute,now 3.6-2.2build1 arm64 [installed]
  Regina REXX interpreter, run-time library

libregina3-dev/hirsute 3.6-2.2build1 arm64
  Regina REXX interpreter, development files

3:
which -a hercules
/home/ubuntu/herctest/herc4x/bin/hercules

Hope this solves the mystery !

Kind regards 

Allard

**Comment by wrljet** [comment]
Allard,

What appears to have happened...

The first hyperion-buildall log you posted from 2021-05-31, in that one, Regina was not already installed on the system,
and was built from source by Hercules-Helper.  The full build process worked successfully.

In your most recent log, 2021-06-13, Regina was already installed, but the development files (.h headers) were missing.
So the build failed, because we told `configure` to expect to find those.

The results you posted today confirm that.

So, at some point between 2021-05-31 and 06-13, you installed Regina REXX using the Ubuntu
package.  But did not install the development files to go with it.  That confused Hercules-Helper's
REXX detection.  (which I will fix by making it stop with an error)

You can get things going again with `sudo apt install libregina3-dev`

Then
```
cd ~/herctest
~/hercules-helper/hyperion-buildall --auto
```

Bill

**Comment by AllardKrings** [comment]
Hi Bill,

Thanks again!

It must have been me, installing Regina Rexx from Debian-package to solve the error, not realising I would confuse the build-script. I removed it on both machines to make sure.

I did as you asked, built the Hercules again on two different Rapspberry-PI's and I believe the script now ran without errors.

If I look in the startup log it seems Hercules is still using Hercifc. I started Hercules as a normal user, no elevated priviliges.

I include the log files:

Kind regards Allard

[hyperion-buildall-2021-06-18.log](https://github.com/SDL-Hercules-390/hyperion/files/6676489/hyperion-buildall-2021-06-18.log)
[herculesstartup.log](https://github.com/SDL-Hercules-390/hyperion/files/6676490/herculesstartup.log)

**Comment by wrljet** [comment]
Allard,

Yes, it is still using hercifc.  The hercules binaries have set-capabilities, so don't require running as root.
All that appears to be working.

There's nothing wrong with installing Regina from the packages.  You just forgot the `libregina3-dev` piece when you did that.
That inconsistent state is something I hadn't anticipated and didn't test for.

Something else I noticed in your last log:  Back up a few comments I suggested that you pull a fresh
copy of Hercules-Helper.  It has many improvements over the one you have.  It appears you are still
using an older one.  And it helps me to get more testing for all the autoconf/configure changes I have
made to Hercules.

Due to some big changes it would be best to delete your hercules-helper directory and get a completely
fresh one.

`git clone https://github.com/wrljet/hercules-helper.git`

Thanks,
Bill
